src/box/engine.h

```diff
@@ -490,6 +501,8 @@ int generic_engine_backup(struct engine *, const struct vclock *,
 void generic_engine_memory_stat(struct engine *, struct engine_memory_stat *);
 void generic_engine_reset_stat(struct engine *);
 int generic_engine_check_space_def(struct space_def *);
+int
+generic_engine_check_space_linearizability(const struct space *space);
```

Comment:
Maybe could be easier to drop all the new virtual functions and simply check `space->engine` + `space->def->opts` manually? Doesn't look like you really need any engine-specific parts of a space. Only its common properties.

Reply:
Good idea, will do tomorrow.

Reply:
Done.
<details><summary>diff</summary>

```diff
diff --git a/src/box/blackhole.c b/src/box/blackhole.c
index 95e0c7e2d..a7857bf0d 100644
--- a/src/box/blackhole.c
+++ b/src/box/blackhole.c
@@ -198,8 +198,6 @@ static const struct engine_vtab blackhole_engine_vtab = {
 	/* .memory_stat = */ generic_engine_memory_stat,
 	/* .reset_stat = */ generic_engine_reset_stat,
 	/* .check_space_def = */ generic_engine_check_space_def,
-	/* .check_space_linearizability = */
-	generic_engine_check_space_linearizability,
 };
 
 struct engine *
diff --git a/src/box/engine.c b/src/box/engine.c
index c47613731..226cd0f77 100644
--- a/src/box/engine.c
+++ b/src/box/engine.c
@@ -461,14 +461,4 @@ generic_engine_check_space_def(struct space_def *def)
 	return 0;
 }
 
-int
-generic_engine_check_space_linearizability(const struct space *space)
-{
-	(void)space;
-	diag_set(ClientError, ER_UNSUPPORTED,
-		 tt_sprintf("space \"%s\"", space_name(space)),
-		 "linearizable operations");
-	return -1;
-}
-
 /* }}} */
diff --git a/src/box/engine.h b/src/box/engine.h
index e985174c4..13d25aaec 100644
--- a/src/box/engine.h
+++ b/src/box/engine.h
@@ -247,10 +247,6 @@ struct engine_vtab {
 	 * tables.
 	 */
 	int (*check_space_def)(struct space_def *);
-	/**
-	 * Check whether the space may be accessed by linearizable requests.
-	 */
-	int (*check_space_linearizability)(const struct space *space);
 };
 
 enum {
@@ -386,13 +382,6 @@ engine_read_view_delete(struct engine_read_view *rv)
 	rv->vtab->free(rv);
 }
 
-static inline int
-engine_check_space_linearizability(struct engine *engine,
-				   const struct space *space)
-{
-	return engine->vtab->check_space_linearizability(space);
-}
-
 /**
  * Shutdown all engine factories.
  */
@@ -501,8 +490,6 @@ int generic_engine_backup(struct engine *, const struct vclock *,
 void generic_engine_memory_stat(struct engine *, struct engine_memory_stat *);
 void generic_engine_reset_stat(struct engine *);
 int generic_engine_check_space_def(struct space_def *);
-int
-generic_engine_check_space_linearizability(const struct space *space);
 
 #if defined(__cplusplus)
 } /* extern "C" */
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 4bb3887f4..8135ae0c9 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -1150,28 +1150,6 @@ memtx_engine_memory_stat(struct engine *engine, struct engine_memory_stat *stat)
 	stat->index += index_stats.totals.used;
 }
 
-/** Check whether the space may be accessed by linearizable operations. */
-static int
-memtx_engine_check_space_linearizability(const struct space *space)
-{
-	/*
-	 * We can't guarantee linearizable access to async spaces. This would
-	 * require checking whether **any** node in the replicaset has
-	 * committed something, which's impossible as soon as at least one node
-	 * becomes unavailable.
-	 * The only exception are local and temporary spaces, which only store
-	 * updates present on this particular node.
-	 */
-	if (!space->def->opts.is_sync && !space->def->opts.is_temporary &&
-	    space->def->opts.group_id != GROUP_LOCAL) {
-		diag_set(ClientError, ER_UNSUPPORTED,
-			 tt_sprintf("space \"%s\"", space_name(space)),
-			 "linearizable operations");
-		return -1;
-	}
-	return 0;
-}
-
 static const struct engine_vtab memtx_engine_vtab = {
 	/* .shutdown = */ memtx_engine_shutdown,
 	/* .create_space = */ memtx_engine_create_space,
@@ -1200,8 +1178,6 @@ static const struct engine_vtab memtx_engine_vtab = {
 	/* .memory_stat = */ memtx_engine_memory_stat,
 	/* .reset_stat = */ generic_engine_reset_stat,
 	/* .check_space_def = */ generic_engine_check_space_def,
-	/* .check_space_linearizability = */
-	memtx_engine_check_space_linearizability,
 };
 
 /**
diff --git a/src/box/service_engine.c b/src/box/service_engine.c
index a505bd6ee..db2fc70aa 100644
--- a/src/box/service_engine.c
+++ b/src/box/service_engine.c
@@ -115,8 +115,6 @@ static const struct engine_vtab service_engine_vtab = {
 	/* .memory_stat = */ generic_engine_memory_stat,
 	/* .reset_stat = */ generic_engine_reset_stat,
 	/* .check_space_def = */ generic_engine_check_space_def,
-	/* .check_space_linearizability = */
-	generic_engine_check_space_linearizability,
 };
 
 struct engine *
diff --git a/src/box/sysview.c b/src/box/sysview.c
index 03bb808d2..112abe96e 100644
--- a/src/box/sysview.c
+++ b/src/box/sysview.c
@@ -587,8 +587,6 @@ static const struct engine_vtab sysview_engine_vtab = {
 	/* .memory_stat = */ generic_engine_memory_stat,
 	/* .reset_stat = */ generic_engine_reset_stat,
 	/* .check_space_def = */ generic_engine_check_space_def,
-	/* .check_space_linearizability = */
-	generic_engine_check_space_linearizability,
 };
 
 struct sysview_engine *
diff --git a/src/box/txn.c b/src/box/txn.c
index 5c8024a74..0af2e5e3b 100644
--- a/src/box/txn.c
+++ b/src/box/txn.c
@@ -608,7 +608,7 @@ txn_begin_stmt(struct txn *txn, struct space *space, uint16_t type)
 	}
 
 	if (space != NULL && txn_has_flag(txn, TXN_IS_LINEARIZABLE) &&
-	    engine_check_space_linearizability(space->engine, space) != 0) {
+	    txn_check_space_linearizability(txn, space) != 0) {
 		return -1;
 	}
 
diff --git a/src/box/txn.h b/src/box/txn.h
index a8a7e4501..2c52c2b8b 100644
--- a/src/box/txn.h
+++ b/src/box/txn.h
@@ -37,6 +37,7 @@
 #include "fiber.h"
 #include "space.h"
 #include "journal.h"
+#include "tt_static.h"
 
 #if defined(__cplusplus)
 extern "C" {
@@ -565,7 +566,7 @@ struct txn {
 };
 
 static inline bool
-txn_has_flag(struct txn *txn, enum txn_flag flag)
+txn_has_flag(const struct txn *txn, enum txn_flag flag)
 {
 	assert((flag & (flag - 1)) == 0);
 	return (txn->flags & flag) != 0;
@@ -782,6 +783,42 @@ txn_begin_stmt(struct txn *txn, struct space *space, uint16_t type);
 int
 txn_begin_in_engine(struct engine *engine, struct txn *txn);
 
+static inline int
+txn_check_space_linearizability(const struct txn *txn,
+				const struct space *space)
+{
+	assert(txn_has_flag(txn, TXN_IS_LINEARIZABLE));
+	(void)txn;
+	/*
+	 * XXX: can't promise the user linearizability with Vinyl until
+	 * https://github.com/tarantool/tarantool/issues/5522 is fixed.
+	 * For now it doesn't matter if txn isolation is set, mvcc is turned on
+	 * or the space is synchronous.
+	 */
+	if (!space_is_memtx(space)) {
+		diag_set(ClientError, ER_UNSUPPORTED,
+			 tt_sprintf("space \"%s\"", space_name(space)),
+			 "linearizable operations");
+		return -1;
+	}
+	/*
+	 * We can't guarantee linearizable access to async spaces. This would
+	 * require checking whether **any** node in the replicaset has
+	 * committed something, which's impossible as soon as at least one node
+	 * becomes unavailable.
+	 * The only exception are local and temporary spaces, which only store
+	 * updates present on this particular node.
+	 */
+	if (!space_is_sync(space) && !space_is_temporary(space) &&
+	    !space_is_local(space)) {
+		diag_set(ClientError, ER_UNSUPPORTED,
+			 tt_sprintf("space \"%s\"", space_name(space)),
+			 "linearizable operations");
+		return -1;
+	}
+	return 0;
+}
+
 /**
  * This is an optimization, which exists to speed up selects
  * in autocommit mode. For such selects, we only need to
@@ -802,8 +839,7 @@ txn_begin_ro_stmt(struct space *space, struct txn **txn,
 			return -1;
 		}
 		if (txn_has_flag(*txn, TXN_IS_LINEARIZABLE) &&
-		    engine_check_space_linearizability(space->engine,
-						       space) != 0) {
+		    txn_check_space_linearizability(*txn, space) != 0) {
 			return -1;
 		}
 		struct engine *engine = space->engine;
diff --git a/src/box/vinyl.c b/src/box/vinyl.c
index 956ae3231..f8a4bb916 100644
--- a/src/box/vinyl.c
+++ b/src/box/vinyl.c
@@ -4537,13 +4537,6 @@ static const struct engine_vtab vinyl_engine_vtab = {
 	/* .reset_stat = */ vinyl_engine_reset_stat,
 	/* .check_space_def = */ vinyl_engine_check_space_def,
 	/* .check_space_linearizability = */
-	/*
-	 * XXX: can't promise the user linearizability with Vinyl until
-	 * https://github.com/tarantool/tarantool/issues/5522 is fixed.
-	 * For now it doesn't matter if txn isolation is set, mvcc is turned on
-	 * or the space is synchronous.
-	 */
-	generic_engine_check_space_linearizability,
 };
 
 static const struct space_vtab vinyl_space_vtab = {
```

</details>